R/tokens_segment.R

```diff
@@ -98,3 +98,66 @@ tokens_segment.tokens <- function(x, pattern,
     return(result)
 
 }
+
+#' Segment tokens object by chunks of a given size
+#' 
+#' Segment tokens by splitting into equally sized chunks.
+#' @param x \link{tokens} object whose token elements will be segmented into
+#'   chunks
+#' @param size integer; the size of the chunks in tokens
+#' @param discard_remainder logical; if \code{TRUE}, discard any ending chunk
+#'   whose length is less than \code{size}
+#' @return A \link{tokens} object whose documents have been split into chunks,
+#'   similar to \code{\link{tokens_segment}}.
+#' @keywords tokens internal
+#' @export
+#' @examples 
+#' txts <- c(doc1 = "Fellow citizens, I am again called upon by the voice of 
+#'                   my country to execute the functions of its Chief Magistrate.",
+#'           doc2 = "When the occasion proper for it shall arrive, I shall 
+#'                   endeavor to express the high sense I entertain of this
+#'                   distinguished honor.")
+#' toks <- tokens(txts)
+#' tokens_chunk(toks, size = 5)
+tokens_chunk <- function(x, size, discard_remainder = TRUE) {
+    UseMethod("tokens_chunk")
+}
+
+#' @export
+tokens_chunk.tokens <- function(x, size, discard_remainder = TRUE) {
```

Comment:
I think I can do this more efficiently by `tokens_segment()`.

Reply:
No doubt! `tokens_chunk()` is already built on `tokens_segment()` after inserting a value on which to segment the text into regular-sized "chunks". It's exactly like `tokens_segment()` but with the segments defined by length, rather than by delimiter. The function to do the insertion is called (unimaginatively, but descriptively) `insert_values()` in `tokens_segment.R`. It's pretty efficient for R. So it may not be worth modifying `tokens_segment()` further at this point for what is probably a very specific use-case.

But then maybe not. I can see where we could certainly want the ability to reshape a corpus or a tokens object into fixed-length texts or tokens, rather than segmenting them on a pattern.

Reply:
We had discussion on this here https://github.com/quanteda/quanteda/issues/679. If it allows overlapping of tokens, there must be use cases (e.g. automated audio transcripts that do not have punctuation at all!).

For example, `list(c("a", "b", "c", "d")` will be
```
list(c("a", "b"), c("c", "d"))
```
when `length = 2` and `overlap = FALSE`, 

```
list(c("a", "b"), c("b", "c"), c("c", "d"))
```
when `length = 2` and `overlap = TRUE`. This also solves the moving window problem.

Reply:
True! I'd forgotten. We also had a lot of discussion about the segment versus reshape options, and went with `corpus_segment()` and `tokens_segment()` operating on a pattern match, and `corpus_reshape()` being limited to redefining documents based on syntactical divisions. `corpus_segment()` does not need to be tokenized first because it looks for a pattern in the text, rather than a pattern match to tokens.

I think it's clear to us that:
- we need a function similar to what in this PR is called `tokens_chunk()`
- the addition of the `overlap` option is an interesting idea, although I am not sure what would be the use case
- `tokens_chunk()` makes sense but `corpus_chunk()` does not, unless it operates only on characters (since it has no "tokens" until it has been tokenized)

I dislike the idea of building a "chunk_size" argument into `tokens_segment()` because it would be incompatible with the pattern argument, which is currently a required argument, and because then it would behave differently from `corpus_segment()` that also splits on patterns.

I'm not exactly wild about `tokens_chunk()` as a name although that's pretty much an accurate description, and I cannot think of anything better right now.

Reply:
Then, we can add `tokens_chunk()` as an experimental function. `overlap` can be used, for example, to detect text reuse:
```r
tokens_chunk(toks, size = 10, overlap = TRUE) %>%
dfm() %>% 
textstat_proxy(min_prox = 0.9)
```